### PR TITLE
p2p: Replace RecursiveMutex `cs_totalBytesSent`  with Mutex and rename it

### PR DESCRIPTION
**Pull request by w0xlt**
Related to #19303, this PR gets rid of the RecursiveMutex `cs_totalBytesSent` and also adds `AssertLockNotHeld` macros combined with `LOCKS_EXCLUDED` thread safety annotations to avoid recursive locking.

### CODE REVIEW

File: src/net.h [hunk 1]
```diff
@@ -910,24 +911,22 @@ class CConnman
     //! that peer during `net_processing.cpp:PushNodeVersion()`.
     ServiceFlags GetLocalServices() const;
 
-    uint64_t GetMaxOutboundTarget() const;
+    uint64_t GetMaxOutboundTarget() const LOCKS_EXCLUDED(m_total_bytes_sent_mutex);
```
Comment (hebasto) [inline src/net.h:914]: Is it feasible
```suggestion
    uint64_t GetMaxOutboundTarget() const EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
```
here and after?

For context and example see #24235?

Comment (vasild) [reply src/net.h:914]: I think yes.

Comment (hebasto) [reply src/net.h]: And in other places?

**Comment by w0xlt** [comment]
Addressed @hebasto suggestion.

All functions changed in the second commit use  `EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex)` instead of `LOCKS_EXCLUDED(m_total_bytes_sent_mutex)`.

The only exception is `void RecordBytesSent(uint64_t bytes)` which displayed the error below when using `EXCLUSIVE_LOCKS_REQUIRED`.

Not sure about the reason for this error.

```c++
net.cpp:1655:29: error: calling function 'RecordBytesSent' requires negative capability '!m_total_bytes_sent_mutex' [-Werror,-Wthread-safety-analysis]
            if (bytes_sent) RecordBytesSent(bytes_sent);
                            ^
net.cpp:3077:21: error: calling function 'RecordBytesSent' requires negative capability '!m_total_bytes_sent_mutex' [-Werror,-Wthread-safety-analysis]
    if (nBytesSent) RecordBytesSent(nBytesSent);

```

**Comment by hebasto** [comment]
@w0xlt 
> Not sure about the reason for this error.

Maybe:
```diff
--- a/src/net.h
+++ b/src/net.h
@@ -829,7 +829,8 @@ public:
 
     bool ForNode(NodeId id, std::function<bool(CNode* pnode)> func);
 
-    void PushMessage(CNode* pnode, CSerializedNetMsg&& msg);
+    void PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
 
     using NodeFn = std::function<void(CNode*)>;
     void ForEachNode(const NodeFn& func)
@@ -1024,7 +1025,8 @@ private:
     /**
      * Check connected and listening sockets for IO readiness and process them accordingly.
      */
-    void SocketHandler();
+    void SocketHandler()
+        EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
 
     /**
      * Do the read/write for connected sockets that are ready for IO.
@@ -1037,7 +1039,8 @@ private:
     void SocketHandlerConnected(const std::vector<CNode*>& nodes,
                                 const std::set<SOCKET>& recv_set,
                                 const std::set<SOCKET>& send_set,
-                                const std::set<SOCKET>& error_set);
+                                const std::set<SOCKET>& error_set)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
 
     /**
      * Accept incoming connections, one from each read-ready listening socket.
@@ -1045,7 +1048,8 @@ private:
      */
     void SocketHandlerListening(const std::set<SOCKET>& recv_set);
 
-    void ThreadSocketHandler();
+    void ThreadSocketHandler()
+        EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
     void ThreadDNSAddressSeed();
 
     uint64_t CalculateKeyedNetGroup(const CAddress& ad) const;
@@ -1074,7 +1078,8 @@ private:
 
     // Network stats
     void RecordBytesRecv(uint64_t bytes);
-    void RecordBytesSent(uint64_t bytes) LOCKS_EXCLUDED(m_total_bytes_sent_mutex);
+    void RecordBytesSent(uint64_t bytes)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
 
     /**
      * Return vector of current BLOCK_RELAY peers.
```
?

**Comment by w0xlt** [comment]
@hebasto Thanks. This worked.

Very interesting. So, using ` EXCLUSIVE_LOCKS_REQUIRED(!mutex)` requires that all functions that call the annotated function also have the same annotation.  It is a cascading effect.

**Comment by hebasto** [comment]
> @hebasto Thanks. This worked.
> 
> Very interesting. So, using ` EXCLUSIVE_LOCKS_REQUIRED(!mutex)` requires that all functions that call the annotated function also have the same annotation. It is a cascading effect.

That is why [negative capabilities](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#negative) provide a stronger safety guarantee.

---

I don't see any reason of splitting changes between second and third commits. Maybe squash them?

**Comment by w0xlt** [comment]
>Maybe squash them?

Sure. Done.

File: src/net.h [hunk 2]
```diff
@@ -1034,15 +1037,17 @@ class CConnman
     void SocketHandlerConnected(const std::vector<CNode*>& nodes,
                                 const std::set<SOCKET>& recv_set,
                                 const std::set<SOCKET>& send_set,
-                                const std::set<SOCKET>& error_set);
+                                const std::set<SOCKET>& error_set)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
+;
```
Comment (hebasto) [inline src/net.h:1042]: A spare semicolon? :smile:

**Review by hebasto — APPROVED** [review]
ACK 0b6c5da9304af71dedb262bca9a0493c8616985b

File: src/net.h (hunk 2 above)
Comment (w0xlt) [reply src/net.h]: Oops. Fixed.

**Review by hebasto — APPROVED** [review]
re-ACK f04ffdd2a79a77ea70d892be972de7821c1b876c

File: src/net.cpp [hunk 3]
```diff
@@ -2927,7 +2929,15 @@ std::chrono::seconds CConnman::GetMaxOutboundTimeframe() const
 
 std::chrono::seconds CConnman::GetMaxOutboundTimeLeftInCycle() const
 {
-    LOCK(cs_totalBytesSent);
+    AssertLockNotHeld(m_total_bytes_sent_mutex);
+    LOCK(m_total_bytes_sent_mutex);
+    return _GetMaxOutboundTimeLeftInCycle();
+}
+
+std::chrono::seconds CConnman::_GetMaxOutboundTimeLeftInCycle() const
```
Comment (vasild) [inline src/net.cpp:2937]: nit: I think that the unwritten convention is to put the underscore at the end, like e.g. `AddrManImpl::Good_()`. Or maybe I am mistaken, is there any other place in the code that uses prefix `_`?

File: src/net.h [hunk 4]
```diff
@@ -796,7 +796,8 @@ class CConnman
         nReceiveFloodSize = connOptions.nReceiveFloodSize;
         m_peer_connect_timeout = std::chrono::seconds{connOptions.m_peer_connect_timeout};
         {
-            LOCK(cs_totalBytesSent);
+            AssertLockNotHeld(m_total_bytes_sent_mutex);
+            LOCK(m_total_bytes_sent_mutex);
```
Comment (vasild) [inline src/net.h:800]: This is not the beginning of a function. `AssertLockNotHeld()` does not belong here - it should be put in the beginning of functions.

File: src/net.h [hunk 5]
```diff
@@ -828,7 +829,7 @@ class CConnman
 
     bool ForNode(NodeId id, std::function<bool(CNode* pnode)> func);
 
-    void PushMessage(CNode* pnode, CSerializedNetMsg&& msg);
+    void PushMessage(CNode* pnode, CSerializedNetMsg&& msg) EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
```
Comment (vasild) [inline src/net.h:832]: Why not add `AssertLockNotHeld(m_total_bytes_sent_mutex)` in the beginning of `PushMessage()`, `SocketHandler()`, `SocketHandlerConnected()` and `ThreadSocketHandler()`?

File: src/net.cpp (hunk 3 above)
Comment (w0xlt) [reply src/net.cpp]: Changed the underscore to the end. I think only in `AddrManImpl` this is used.

File: src/net.h (hunk 4 above)
Comment (w0xlt) [reply src/net.h]: Fixed. Thanks.

File: src/net.h (hunk 5 above)
Comment (w0xlt) [reply src/net.h:833]: Good catch. Done. Thanks.

**Review by vasild — APPROVED** [review]
ACK eef2c7b8c2d3c35da027429bf89ebb792e55f739

**Review by hebasto — APPROVED** [review]
~re-ACK eef2c7b8c2d3c35da027429bf89ebb792e55f739~, only suggested changes since my previous [review](https://github.com/bitcoin/bitcoin/pull/24157#pullrequestreview-881048320).

UPDATE:  ACK retracted. See https://github.com/bitcoin/bitcoin/pull/24157#discussion_r849775212

**Comment by w0xlt** [comment]
Is any further action needed in this PR?

File: src/net.h [hunk 6]
```diff
@@ -781,6 +781,8 @@ class CConnman
     };
 
     void Init(const Options& connOptions) {
```
Comment (ajtowns) [inline src/net.h:783]: Missing `EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex)`

**Review by ajtowns — COMMENTED** [review]
Think there's a missing annotation, but otherwise looks good.

File: src/net.h (hunk 6 above)
Comment (hebasto) [reply src/net.h:783]: Agree. I've verified this PR rebased on top of the current master (f60a63cc5f16b738d9d2ada3f10b27cf999df323):
```
$ ./configure CC=clang CXX=clang++ CXXFLAGS='-Wthread-safety -Wthread-safety-negative'
$ make clean
$ make 2>&1 | grep m_total_bytes_sent_mutex
```
ends with warnings
```
./net.h:781:13: warning: acquiring mutex 'm_total_bytes_sent_mutex' requires negative capability '!m_total_bytes_sent_mutex' [-Wthread-safety-negative]
            LOCK(m_total_bytes_sent_mutex);
```

To silent all of them, the following patch works:
```diff
--- a/src/net.h
+++ b/src/net.h
@@ -760,7 +760,8 @@ public:
         bool m_i2p_accept_incoming;
     };
 
-    void Init(const Options& connOptions) {
+    void Init(const Options& connOptions) EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex)
+    {
         AssertLockNotHeld(m_total_bytes_sent_mutex);
 
         nLocalServices = connOptions.nLocalServices;
@@ -791,7 +792,7 @@ public:
 
     CConnman(uint64_t seed0, uint64_t seed1, AddrMan& addrman, bool network_active = true);
     ~CConnman();
-    bool Start(CScheduler& scheduler, const Options& options);
+    bool Start(CScheduler& scheduler, const Options& options) EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
 
     void StopThreads();
     void StopNodes();
```

**Comment by jonatack** [comment]
Concept ACK, will review tomorrow.

File: src/net.h (hunk 6 above)
Comment (jonatack) [reply src/net.h:783]: Reproduced the issue and confirm that this patch works.

**Comment by jonatack** [comment]
ACK modulo the missing annotation

Verified the methods updated here have both an annotation in the declaration and a corresponding assertion in the definition:
```
GetMaxOutboundTarget
GetMaxOutboundTimeLeftInCycle
GetMaxOutboundTimeLeftInCycle_
GetOutboundTargetBytesLeft
GetTotalBytesSent
OutboundTargetReached
PushMessage
RecordBytesSent
SocketHandler
SocketHandlerConnected
ThreadSocketHandler
```

File: src/net.h (hunk 6 above)
Comment (w0xlt) [reply src/net.h]: Done. Thanks for the reviews and suggestions.

File: src/net.h [hunk 7]
```diff
@@ -789,7 +792,7 @@ class CConnman
 
     CConnman(uint64_t seed0, uint64_t seed1, AddrMan& addrman, bool network_active = true);
     ~CConnman();
-    bool Start(CScheduler& scheduler, const Options& options);
+    bool Start(CScheduler& scheduler, const Options& options) EXCLUSIVE_LOCKS_REQUIRED(!m_total_bytes_sent_mutex);
```
Comment (jonatack) [inline src/net.h:795]: e59472f8af0a1962fa5c331e41fa22a4f8602db8 In the corresponding definition in `src/net.cpp`
```diff
 bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
 {
+    AssertLockNotHeld(m_total_bytes_sent_mutex);
     Init(connOptions);
```

**Review by vasild — APPROVED** [review]
ACK eff791863da5dd1fcda7bddc7d729a63d772659d

Compiles with clang 15.

File: src/net.h (hunk 7 above)
Comment (w0xlt) [reply src/net.h:795]: Done. Thanks.

**Comment by w0xlt** [comment]
Added the suggestion https://github.com/bitcoin/bitcoin/pull/24157#discussion_r852269547

**Review by vasild — APPROVED** [review]
ACK 709af67add93f6674fb80e3ae8e3f175653a62f0

**Review by hebasto — APPROVED** [review]
ACK 709af67add93f6674fb80e3ae8e3f175653a62f0, tested on Ubuntu 22.04.

**Comment by jonatack** [comment]
ACK 709af67add93f6674fb80e3ae8e3f175653a62f0 per `git range-diff 7a4ac71 eff7918 709af67`, rebase to master, clang 15 debug build, and build with -Wthread-safety-negative

**Pull request merged**